streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java

```diff
@@ -317,14 +318,31 @@ public boolean isOpen() {
     @Override
     public <R> QueryResult<R> query(final Query<R> query, final PositionBound positionBound,
         final boolean collectExecutionInfo) {
+
+        if (query instanceof WindowRangeQuery) {
+            @SuppressWarnings("unchecked") final WindowRangeQuery<Bytes, byte[]> windowRangeQuery = (WindowRangeQuery<Bytes, byte[]>) query;
+            if (windowRangeQuery.getKey().isPresent()) {
+                final Bytes key = windowRangeQuery.getKey().get();
+                final KeyValueIterator<Windowed<Bytes>, byte[]> keyValueIterator = this.fetch(key);
+                @SuppressWarnings("unchecked") final R result = (R) keyValueIterator;
+                final QueryResult<R> queryResult = QueryResult.forResult(result);
+                return queryResult;
+            }
+        }
+
+        return null;
+
+        /*
         return StoreQueryUtils.handleBasicQueries(
-            query,
-            positionBound,
-            collectExecutionInfo,
-            this,
-            position,
-            context.taskId().partition()
+                query,
+                positionBound,
+                collectExecutionInfo,
+                this,
+                position,
+                context.taskId().partition()
         );
+
+         */
```

Comment:
I assume that the commented code is just a temporary state while you're finalizing the PR.

Since executing these queries doesn't require any specialization to the concrete store type, just the WindowStore and SessionStore interfaces, I think you might want to move the logic into StoreQueryUtils, but I've been thinking that, rather than having a universal method to `handleBasicQueries` on any state store, the code will be cleaner with less type-checks and casts if you instead add a new `handleWindowQueries` and `handleSessionQueries`.

What do you think?

Reply:
I was looking at your merged PR and coded the store query routines in a similar spirit. For instance, I follow your practice in the MeteredKeyValue store and applied it to MeteredWindow/Session stores. Similarly for the InMemoryKeyValue store I noticed you handle range queries in place, so I did the same for the InMemoryWindowStore. Generally if you feel we should handle queries in the util class (as some are) then probably we should be consistent across the stores and remove the local handlers and instead delegate to the store utils?